feat: add and check against Speed Mode at SFX confirmation [skip-release]

MaciejBaj opened the pull request:
# Summary of changes

feat: add and check against Speed Mode at SFX confirmation

!breaks! client packages submission on_extrinsic_trigger since sequential is now replaced with SpeedMode

# Acceptance Checklist

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

## Dependencies

- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Any teams which manage the dependencies have been notified of the breaking changes

## Bug (non-breaking change which fixes an issue):

- [x] Have you **_written tests_** to protect against future occurrences of the bug?

## Feature (non-breaking change which adds functionality)

- [x] Have you **_seen it working_** in a development environment?
- [ ] Have you **_written tests_** for **_happy_** and **_unhappy_** paths?
- [ ] Have you implemented this feature as per the **_issue acceptance criteria_**?

### Substrate

If this change involves substrate, have you remembered:

- [ ] Storage Migration?
- [ ] RPC methods?
- [x] Chainspec, both JSON specs & the module?
- [ ] Runtime versioning?
- [ ] Benchmarks?

## Breaking change (a feature that would cause existing functionality not to work as expected)

- [ ] Is the breaking change **_documented_**?
- [ ] Are your commits fully representative of the change?
- [ ] Has any previous functionality been **_deprecated_** and versioned?

## CI/CD

- [ ] If introducing new actions, have you **_looked at the action code_** for anything spurious?
- [ ] Have you written **_custom scripts_** for things that could be actions already on the marketplace?
- [ ] If introducing new actions, have you **_pegged a static version_**?

## Documentation Update

- [ ] Have you checked other documentation, such as the docs repository?
- [ ] If updating script documentation, have you **_tested it on both environments_**?

## Further comments

Feel free to explain in further detail your choices if this is a significant change.

## Screenshots (Please demonstrate this working in PolkadotJS)

AwesomeIbex commented on pallets/attesters/Cargo.toml:
```diff
@@ -0,0 +1,59 @@
+[package]
+name        = "pallet-attesters"
+edition     = "2018"
```
```suggestion
edition     = "2021"
```

AwesomeIbex commented on pallets/attesters/Cargo.toml:
```diff
@@ -0,0 +1,59 @@
+
+log              = { version = "0.4", default-features = false }
+
+sp-core            = { git = "https://github.com/paritytech/substrate.git", branch = 'polkadot-v0.9.27', default-features = false}
```
format this pls. Do `taplo format` from root of t3rn

AwesomeIbex commented on pallets/attesters/src/lib.rs:
```diff
@@ -0,0 +1,864 @@
+    }
+
+    #[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+    pub struct AttesterInfo {
+        pub key_ed: [u8; 32],
+        pub key_ec: [u8; 33],
+        pub key_sr: [u8; 32],
+    }
```
I think this can be better represented as an enum which is:

- `From<KeyTypeId>`
- `AsRef` for into `&[u8]`
- `Self::verify` as a function
- #[default] over ECDSA since we don't use the others right now
- `.0` as the pkey bytes

AwesomeIbex commented on pallets/circuit/src/lib.rs:
```diff
@@ -552,12 +553,14 @@ pub mod pallet {
             origin: OriginFor<T>,
             side_effects: Vec<SideEffect<T::AccountId, BalanceOf<T>>>,
-            _sequential: bool,
+            speed_mode: SpeedMode,
```
will there always be a speed mode? If we made this an option we can pass None and it will use a sensible default.

AwesomeIbex commented on pallets/circuit/src/lib.rs:
```diff
@@ -627,7 +630,7 @@ pub mod pallet {
                 |current_fsx, _local_state, _steps_cnt, __status, _requester| {
                     Self::confirm(xtx_id, current_fsx, &sfx_id, &confirmation).map_err(|e| {
-                        log::error!("Self::confirm hit an error -- {:?}", e);
+                        println!("Self::confirm hit an error -- {:?}", e);
```
put this back please

AwesomeIbex commented on pallets/circuit/src/lib.rs:
```diff
@@ -996,89 +999,108 @@ impl<T: Config> Pallet<T> {
+        let target = fsx.input.target;
+        let xtx = &mut Machine::<T>::load_xtx(xtx_id)?.xtx;
+
+        match xtx.speed_mode {
+            SpeedMode::Fast => {
+                if fsx.submission_target_height + T::Portal::read_fast_confirmation_offset(target)?
+                    < current_target_height
+                {
+                    return Err(DispatchError::Other(
+                        "SFX validate failed - fast confirmation offset not satisfied",
+                    ))
+                }
+            },
+            SpeedMode::Rational => {
+                if fsx.submission_target_height
+                    + T::Portal::read_rational_confirmation_offset(target)?
+                    < current_target_height
+                {
+                    return Err(DispatchError::Other(
+                        "SFX validate failed - slow confirmation offset not satisfied",
+                    ))
+                }
+            },
+            SpeedMode::Finalized => {
+                if fsx.submission_target_height < current_finalized_target_height {
+                    return Err(DispatchError::Other(
+                        "SFX validate failed - finalized confirmation offset not satisfied",
+                    ))
+                }
+            },
+        }
```
I think this should be extracted to a unit test if not already

AwesomeIbex commented on pallets/evm/Cargo.toml:
```diff
@@ -87,6 +89,7 @@ std = [
   "frame-support/std",
   "frame-system/std",
   "pallet-timestamp/std",
+  "t3rn-abi/std",
```
will also want try-runtime here

AwesomeIbex commented on pallets/evm/precompile/portal/src/mock.rs:
```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: Apache-2.0
```
We are probably gonna delete this now ya?

AwesomeIbex commented on pallets/evm/precompile/util/src/lib.rs:
```diff
@@ -26,6 +28,21 @@ pub enum KnownPrecompile {
     // 3VM precompiles:
 }
 
+pub enum CustomPrecompile<T, BlockNumber> {
```
All of this portal precompile stuff will need a rebase

AwesomeIbex commented on runtime/mock/src/contracts_config.rs:
```diff
@@ -135,10 +137,138 @@ impl FeeCalculator for FixedGasPrice {
     }
 }
 
+impl PortalReadApi<BlockNumber> for Runtime {
```
Also not needed with new stuff, willl omit any more comments from the previous precompile PR from here on

AwesomeIbex commented:
Also please do clippy suggestions

AwesomeIbex commented on client/packages/cli/package-lock.json:
```diff
@@ -0,0 +1,4919 @@
+{
```
delete this pls, we use yarn

MaciejBaj replied:
deleted

AwesomeIbex commented on pallets/circuit/src/lib.rs:
```diff
@@ -996,99 +1000,86 @@ impl<T: Config> Pallet<T> {
             Some(fsx.submission_target_height), // this enforces the submission height check!
         )
-        .map_err(|_| "SideEffect confirmation of inclusion failed")?;
+        .map_err(|_| DispatchError::Other("SideEffect confirmation of inclusion failed"))?;
```
fyi you can get away with "dsjfsfdjkh".into() here, no need to change

AwesomeIbex commented:
FYI you havent filled out the checkboxes regarding dependencies ;)

AwesomeIbex commented on client/packages/sdk/src/circuit/index.ts:
```diff
@@ -6,6 +6,12 @@ import { Tx } from "./tx";
 const DECIMALS = 12;
 const VALUE_TYPE_SIZE = 16;
 
+export enum SpeedMode {
```
We dont get this from t3rn types already?

AwesomeIbex commented on primitives/src/circuit.rs:
```diff
@@ -1,7 +1,9 @@
 use crate::xtx::LocalState;
+pub use crate::SpeedMode;
```
```suggestion
use crate::SpeedMode;
```

People can get this from crate anyway

AwesomeIbex commented on primitives/src/light_client.rs:
```diff
@@ -1,3 +1,95 @@
+use frame_system::pallet_prelude::OriginFor;
+use scale_info::TypeInfo;
+use sp_runtime::DispatchError;
+use t3rn_abi::types::Bytes;
```
```suggestion
use t3rn_abi::types::Bytes;

```

AwesomeIbex reviewed (changes requested):
Some nitpicks, some comments.

Ideally we move the speed check out of circuit and into event inclusion verification.

Please fix all the clippy lints and the coverage patch is only at 53%. We need at least 70% for this PR.

If it is saying to test mocks, check the coverage.yml in the project root to ensure that **mock.rs is excluded (it should be iirc).

MaciejBaj commented on pallets/circuit/Cargo.toml:
```diff
@@ -60,7 +60,7 @@ circuit-runtime-pallets = { path = "../../runtime/common-pallets" }
 
 [features]
-default = [ "std" ]
+default = [ "std", "test-skip-verification" ]
```
removed

MaciejBaj commented:
Fixed the client submission and updated client/packages/Makefile commands. 

Tested with executors + rangers + explorer 

<img width="1134" alt="Screenshot 2023-04-28 at 18 09 05" src="https://user-images.githubusercontent.com/6909553/235199173-8f0a01fb-deb3-48b7-820c-63432f6f3a51.png">
<img width="722" alt="Screenshot 2023-04-28 at 18 09 21" src="https://user-images.githubusercontent.com/6909553/235199179-5a99d04b-b16c-4eb0-9717-54ba501edea0.png">
<img width="726" alt="Screenshot 2023-04-28 at 18 09 30" src="https://user-images.githubusercontent.com/6909553/235199184-b31a0a7e-1644-4756-b527-820e3c27780a.png">
